nemo/collections/nlp/data/language_modeling/megatron/ul2_dataset.py

```diff
@@ -0,0 +1,247 @@
+        num_epochs,
+        max_num_samples,
+        max_seq_length,
+        max_seq_length_dec,
+        seed,
+        masked_lm_prob=0.15,
+        extreme_masked_lm_prob=0.5,
+        short_seq_prob=0.0,
+        min_ngram_size=2,
+        max_ngram_size=10,
+        mean_ngram_size=3,
+        extreme_max_ngram_size=128,
+        extreme_min_ngram_size=32,
+        extreme_mean_ngram_size=64,
+        prefix_lm_pivot_mean=0.25,  # This is represented as a percentage of the total length.
+        ngram_span_length_distribution="geometric",
+        extreme_ngram_span_length_distribution="truncated_normal",
+        permutation=False,
+        whole_word_masking=True,
+        favor_long_ngrams=False,
+    ):
+        super().__init__(
+            cfg=cfg,
+            trainer=trainer,
+            tokenizer=tokenizer,
+            name=name,
+            indexed_dataset=indexed_dataset,
+            data_prefix=data_prefix,
+            num_epochs=num_epochs,
+            max_num_samples=max_num_samples,
+            max_seq_length=max_seq_length - 1,  # -1 to account for the added mask type token
+            max_seq_length_dec=max_seq_length_dec,
+            seed=seed,
+            masked_lm_prob=masked_lm_prob,
+            short_seq_prob=short_seq_prob,
+            max_ngram_size=max_ngram_size,
+            mean_ngram_size=None,  # TODO: Determin if we want to actually pass mean ngram as an override to max here.
+            geometric_dist=ngram_span_length_distribution == "geometric",
+            permutation=permutation,
+            whole_word_masking=whole_word_masking,
+            favor_long_ngrams=favor_long_ngrams,
+        )
+        self.mean_ngram_size = mean_ngram_size
+        self.min_ngram_size = min_ngram_size
+        self.extreme_masked_lm_prob = extreme_masked_lm_prob
+        self.extreme_min_ngram_size = extreme_min_ngram_size
+        self.extreme_max_ngram_size = extreme_max_ngram_size
+        self.extreme_mean_ngram_size = extreme_mean_ngram_size
+        self.ngram_span_length_distribution = ngram_span_length_distribution
+        self.extreme_ngram_span_length_distribution = extreme_ngram_span_length_distribution
+        self.prefix_lm_pivot_mean = prefix_lm_pivot_mean
+        if ngram_span_length_distribution not in ['truncated_normal', 'geometric', 'uniform']:
+            raise ValueError(
+                f'Invalid ngram_span_length_distribution: {ngram_span_length_distribution}. Options : [truncated_normal, geometric, uniform]'
+            )
+
+    def __getitem__(self, idx):
+
+        start_index, end_index, seq_length = self.samples_mapping[idx]
+        sample = []
+        for index in range(start_index, end_index):
+            sample.append(self.indexed_dataset[index])
+        # Note that this rng state should be numpy and not python since
+        # python randint is inclusive whereas the numpy one is exclusive.
+        np_rng = np.random.RandomState(seed=(self.seed + idx))
```

Comment:
Each idx has the fixed randomness.  Validation datapoints all have the same masks throughout the training. Does it make sense to have random masks for the same data points at epochs?

Reply:
This is a good point. T5 Dataset does it this way as well. I think it is good to have a fixed data order/masking to allow skipping batches if the loss explodes? However, we don't have epoch-level concerns because `build_samples_mapping` will pre-allocate all indexes and replicate samples if needed to create "1 epoch" and so the same example will have different masks across epochs.